File: easy/BinaryTreeInorderTraversal[94]/BinaryTreeInorderTraversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        res = []

        def traverse(root):
            if not root:
                return []
            left = root.left
            right = root.right
            traverse(left)
            res.append(root.val)
            traverse(right)

        traverse(root)
        return res

File: easy/BinaryTreeInorderTraversal[94]/BinaryTreeInorderTraversal.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }

use std::cell::RefCell;
use std::rc::Rc;
impl Solution {
    pub fn inorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut res = vec![];
        fn traverse(root: Option<Rc<RefCell<TreeNode>>>, res: &mut Vec<i32>) {
            if let Some(node) = root {
                let r = node.clone();
                let left = r.borrow().left.clone();
                let right = r.borrow().right.clone();
                traverse(left, res);
                res.push(r.borrow().val.clone());
                traverse(right, res);
            }
        }

        traverse(root, &mut res);
        res
    }
}
